packages/gas-fee-controller/src/GasFeeControllerPolling.ts

```diff
@@ -0,0 +1,97 @@
+import type { RestrictedControllerMessenger } from '@metamask/base-controller';
+import { BaseControllerV2 } from '@metamask/base-controller';
+import { safelyExecute } from '@metamask/controller-utils';
+import type { NetworkClientId } from '@metamask/network-controller';
+import type { Json } from '@metamask/utils';
+import { v1 as random } from 'uuid';
+
+export type PollingCompleteType<N extends string> = {
+  type: `${N}:pollingComplete`;
+  payload: [string];
+};
+
+export default abstract class GasFeeControllerPolling<
+  N extends string,
+  S extends Record<string, Json>,
+  messenger extends RestrictedControllerMessenger<
+    N,
+    any,
+    PollingCompleteType<N> | any,
+    string,
+    string
+  >,
+> extends BaseControllerV2<N, S, messenger> {
+  private readonly intervalLength = 1000;
+
+  private readonly networkClientIdTokensMap: Map<NetworkClientId, Set<string>> =
+    new Map();
+
+  private readonly intervalIds: Record<NetworkClientId, NodeJS.Timeout> = {};
+
+  start(networkClientId: NetworkClientId) {
+    const innerPollToken = random();
+    if (this.networkClientIdTokensMap.has(networkClientId)) {
+      //
+      const set = this.networkClientIdTokensMap.get(networkClientId);
+      set?.add(innerPollToken);
+    } else {
+      const set = new Set<string>();
+      set.add(innerPollToken);
+      this.networkClientIdTokensMap.set(networkClientId, set);
+    }
+    this.#poll(networkClientId);
+
+    // call _poll
+    // add the inner poll token to the poll tokens set
+    return innerPollToken;
+  }
+
+  stop({
+    pollingToken,
+    networkClientId,
+  }: {
+    pollingToken?: string;
+    networkClientId?: NetworkClientId;
+  } = {}) {
+    if (pollingToken && !networkClientId) {
+      throw new Error(
+        'networkClientId is required when pollingToken is passed',
+      );
+    }
+    if (!pollingToken || !networkClientId) {
+      this.networkClientIdTokensMap.forEach((tokens, _networkClientId) => {
+        tokens.forEach((token) => {
+          this.stop({
+            pollingToken: token,
+            networkClientId: _networkClientId,
+          });
+        });
+      });
+      return;
+    }
+    this.networkClientIdTokensMap.get(networkClientId)?.delete(pollingToken);
+    if (this.networkClientIdTokensMap.get(networkClientId)?.size === 0) {
+      clearInterval(this.intervalIds[networkClientId]);
+      delete this.intervalIds[networkClientId];
+      this.networkClientIdTokensMap.delete(networkClientId);
+      this.messagingSystem.publish(
+        `${this.name}:pollingComplete`,
+        networkClientId,
+      );
+    }
+  }
+
+  abstract executePoll(networkClientId: NetworkClientId): Promise<void>;
+
+  #poll(networkClientId: NetworkClientId) {
+    if (this.intervalIds[networkClientId]) {
+      clearInterval(this.intervalIds[networkClientId]);
+    }
+
+    this.intervalIds[networkClientId] = setInterval(async () => {
```

Comment:
Should each networkClientId get its own polling / setInterval? Or should we have 1 single interval, each time it occurs we call each of the `executePoll` methods?

Reply:
While I see that would be a slightly cleaner to not have another collection per networkClientId, we were thinking it might actually be better for performance to have the network requests staggered.

Reply:
I like @BelfordZ 's thinking here of keeping everything on the same "heartbeat". I'm not saying we do this, but we could use the same interval + random wait.

This also reminds me, is setInterval the right approach, or would we be using setTimeout?

Reply:
I suppose it could worth testing the performance tradeoffs. What, if anything, is the basis for this preference other than a cleaner abstraction?

Reply:
Might be better to use `setTimeout` over `setInterval to avoid overlapping queries

See https://developer.mozilla.org/en-US/docs/Web/API/setInterval#ensure_that_execution_duration_is_shorter_than_interval_frequency

Reply:
done here: https://github.com/MetaMask/core/pull/1673/commits/5317aca24f4acee8bffe8f9c1f695592d4a1566b